Solutions/Workshops/workshop3.lean

```diff
@@ -0,0 +1,254 @@
+  have ynegorpos : 0 ≤ y ∨ 0 > y := le_or_gt 0 y
+  rcases ynegorpos with yneg | ypos
+  · rw [abs_of_nonneg yneg]
+    intro h; left; exact h -- The `;` allows us to write tactics on the same line.
+  . rw [abs_of_neg ypos]
+    intro h; right; exact h
+
+-- You can nest patterns in `rcases` and `rintro`, even mixing `|` and `⟨⟩`.
+-- Recall that the vertical bar symbol for divisibility is typed with `\|` but for disjunction or absolute value it is just `|`.
+example {m n k : ℕ} (h : m ∣ n ∨ m ∣ k) : m ∣ n * k := by
+  rcases h with ⟨a, rfl⟩ | ⟨b, rfl⟩
+  -- Recall that using `rfl` in a pattern will rewrite the equation it replaces everywhere.
+  -- Here the first `rfl` will replace `n` with `m * a`, and second will replace `k` with `m * b`.
+  · rw [mul_assoc]
+    apply dvd_mul_right
+  . rw [mul_comm, mul_assoc]
+    apply dvd_mul_right
+
+
+-- You can use this to split into cases, depending on whether a proposition is true or false.
+-- To do this explicitly, you can use the law of the excluded middle:
+#check em
+
+example (P : Prop) : ¬¬P → P := by
+  intro h
+  rcases em P with hp | hnp
+  · assumption
+  . contradiction -- This tactic looks for a contradiction in the hypotheses, in this case `h` and `hnp`.
+
+-- You can also do this using the `by_cases` tactic.
+example (P : Prop) : ¬¬P → P := by
+  intro h
+  by_cases h' : P
+  -- `P` is the proposition we are splitting on, and `h'` is the name of the hypothesis we get.
+  · assumption
+  . contradiction
+
+
+/-!
+  # Exercises
+-/
+
+section
+variable {x y : ℝ}
+
+-- 3. Prove the triangle inequality. [MIL 3.5]
+
+-- You may want to use the following lemmas:
+#check le_abs_self x
+#check neg_le_abs x
+
+example : |x + y| ≤ |x| + |y| := by
+  rw [abs_le']
+  simp only [neg_add_rev, add_neg_le_iff_le_add]
+  exact ⟨add_le_add (le_abs_self _) (le_abs_self _), by linarith [neg_le_abs x, neg_le_abs y]⟩
+
+
+-- 4. Prove the following lemma. [MIL 3.5]
+
+-- You may want to use the following lemma:
+#check eq_zero_or_eq_zero_of_mul_eq_zero
+
+example {x : ℝ} (h : x ^ 2 = 1) : x = 1 ∨ x = -1 := by
+  rw [← sub_eq_zero, ← @sub_eq_zero _ _ x, sub_neg_eq_add]
+  apply eq_zero_or_eq_zero_of_mul_eq_zero
+  ring_nf
+  rw [h, neg_add_self]
+
+end
+
+
+-- 5. Prove the following equivalence, using `by_cases` in one direction [MIL 3.5]
+
+example (P Q : Prop) : P → Q ↔ ¬P ∨ Q := by
+  constructor
+  . intro h
+    by_cases h' : P
+    · right; exact h h'
+    · left; exact h'
+  . rintro (h | h) hp
+    · contradiction
+    · exact h
+
+
+-- 6. Here we define what it means for a function `ℝ → ℝ` to be nondecreasing.
+-- Prove the following lemma [Tut 3]
+
+def NonDecreasing (f : ℝ → ℝ) : Prop :=
+  ∀ x₁ x₂, x₁ ≤ x₂ → f x₁ ≤ f x₂
+
+-- You may want to use:
+#check le_total
+
+example (f : ℝ → ℝ) (h : NonDecreasing f) (h' : ∀ x, f (f x) = x) : ∀ x, f x = x := by
+  intro x
+  rcases le_total (f x) x with H | H
+  · apply le_antisymm H
+    nth_rw 1 [← h' x]
```

Comment:
Should we add a comment about `nth_rw`? I believe we briefly mentioned it, but never actually gave an example.

Reply:
Maybe we should add a file with all the tactics and examples of how to use them? 🤔